arrow/src/array/equal/structure.rs

```diff
@@ -84,7 +63,7 @@ pub(super) fn struct_equal(
 
             lhs_is_null
                 || (lhs_is_null == rhs_is_null)
-                    && equal_values(lhs, rhs, lhs_nulls, rhs_nulls, lhs_pos, rhs_pos, 1)
+                    && equal_child_values(lhs, rhs, lhs_pos, rhs_pos, 1)
```

Comment:
I think this needs the same logic as for list arrays:

```suggestion
            if lhs_is_null != rhs_is_null {
                return false;
            }

            lhs_is_null || equal_child_values(lhs, rhs, lhs_pos, rhs_pos, 1)
```